File: 2058.find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
#
# @lc app=leetcode id=2058 lang=java
#
# [2058] Find the Minimum and Maximum Number of Nodes Between Critical Points
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        if (head == null || head.next == null) {
            return new int[]{-1, -1};
        }
        ListNode p = head;
        ListNode c = head.next;
        int idx = 2;
        int minDist = Integer.MAX_VALUE;
        int firstCrit = -1;
        int prevCrit = -1;
        int lastCrit = -1;
        while (c != null && c.next != null) {
            ListNode n = c.next;
            boolean isCrit = (c.val > p.val && c.val > n.val) || (c.val < p.val && c.val < n.val);
            if (isCrit) {
                if (prevCrit != -1) {
                    minDist = Math.min(minDist, idx - prevCrit);
                }
                prevCrit = idx;
                if (firstCrit == -1) {
                    firstCrit = idx;
                }
                lastCrit = idx;
            }
            p = c;
            c = n;
            idx++;
        }
        if (firstCrit == -1 || firstCrit == prevCrit) {
            return new int[]{-1, -1};
        }
        return new int[]{minDist, lastCrit - firstCrit};
    }
}
# @lc code=end

File: 2058.find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
#
# @lc app=leetcode id=2058 lang=cpp
#
# [2058] Find the Minimum and Maximum Number of Nodes Between Critical Points
#
# @lc code=start
/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode() : val(0), next(nullptr) {}
*     ListNode(int x) : val(x), next(nullptr) {}
*     ListNode(int x, ListNode *next) : val(x), next(next) {}
* };
*/
#include <vector>
#include <climits>
using namespace std;
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> criticalIndices;
        int idx = 1; // start at 1 for the second node (1-based)
        ListNode* prev = head;
        ListNode* curr = head ? head->next : nullptr;
        while (curr && curr->next) {
            ++idx;
            int prevVal = prev->val;
            int currVal = curr->val;
            int nextVal = curr->next->val;
            if ((currVal > prevVal && currVal > nextVal) || (currVal < prevVal && currVal < nextVal)) {
                criticalIndices.push_back(idx - 1); // zero-based index for the current node
            }
            prev = curr;
            curr = curr->next;
        }
        // Verification: ensure at least two critical points
        if (criticalIndices.size() < 2) {
            return {-1, -1};
        }
        int minDist = INT_MAX;
        for (size_t i = 1; i < criticalIndices.size(); ++i) {
            minDist = min(minDist, criticalIndices[i] - criticalIndices[i-1]);
        }
        int maxDist = criticalIndices.back() - criticalIndices.front();
        return {minDist, maxDist};
    }
};
# @lc code=end